### PR TITLE
Intelligent errors.Cause()

### PR DESCRIPTION
**Pull request by jaekwon**
When causer.Cause() returns nil or itself, it is the cause.

Ref: https://github.com/pkg/errors/issues/89

### CODE REVIEW

File: errors.go [hunk 1]
```diff
@@ -261,9 +261,12 @@ func Cause(err error) error {
...
 		cause, ok := err.(causer)
 		if !ok {
-			break
+			return err
+		}
+		errCause := cause.Cause()
+		if errCause == nil || errCause == err {
+			return err
 		}
-		err = cause.Cause()
```
Comment (puellanivis) [inline errors.go]: If you never update err, then this for loop can never finish, and it is now possible that the loop becomes a forever loop.

**Review by puellanivis — CHANGES_REQUESTED** [review]

File: errors.go (hunk 1 above)
Comment (jaekwon) [reply errors.go]: heh, was just about to test.  thank you.

Comment (jaekwon) [reply errors.go]: Done.

**Comment by jaekwon** [comment]
Closing this to open a PR on pkg/errors

**Pull request closed without merging**